src/sources/util/tcp.rs

```diff
@@ -98,117 +94,133 @@ pub trait TcpSource: Clone + Send + 'static {
+        },
+        _ = &mut shutdown => {
+            return;
+        }
+    };
+
     let mut _token = None;
-    let socket = Compat01As03::new(socket).compat();
-    let mut reader = Compat::new(FramedRead::new(socket, source.decoder()));
-    let handler = stream::poll_fn(move || {
-        // Gracefull shutdown procedure
-        if let Some(future) = shutdown.as_mut() {
-            match future.poll() {
-                Ok(Async::Ready(tk)) => {
+    let mut shutdown = Some(shutdown);
+    let mut reader = FramedRead::new(socket, source.decoder());
+    stream::poll_fn(move |cx| {
+        if let Some(fut) = shutdown.as_mut() {
+            match Pin::new(fut).poll(cx) {
+                Poll::Ready(Ok(token)) => {
                     debug!("Start gracefull shutdown");
                     // Close our write part of TCP socket to signal the other side
                     // that it should stop writing and close the channel.
-                    let socket: Option<&TcpStream> = reader.get_ref().get_ref().get_ref().get_ref().get_ref();
+                    let socket: Option<&TcpStream> = reader.get_ref().get_ref();
                     if let Some(socket) = socket {
-                        if let Err(error) = socket.shutdown(Shutdown::Write) {
-                            warn!(message = "Failed in signalling to the other side to close the TCP channel.",%error);
+                        if let Err(error) = socket.shutdown(std::net::Shutdown::Write) {
+                            warn!(message = "Failed in signalling to the other side to close the TCP channel.", %error);
                         }
                     } else {
                         // Connection hasn't yet been established so we are done here.
                         debug!("Closing connection that hasn't yet been fully established.");
-                        return Ok(Async::Ready(None));
+                        return Poll::Ready(None);
                     }
-                    _token = Some(tk);
+
+                    _token = Some(token);
                     shutdown = None;
                 }
-                Err(()) => shutdown = None,
-                Ok(Async::NotReady) => (),
+                Poll::Ready(Err(())) => {
+                    shutdown = None;
+                }
+                Poll::Pending => {}
             }
         }
 
-        // Actual work
-        reader.poll()
+        Pin::new(&mut reader).poll_next(cx)
     })
     .take_until(tripwire)
-    .filter_map(move |frame| {
-        let host = host.clone();
-        source.build_event(frame, host)
-    })
-    .map_err(|error| {
-        emit!(TcpConnectionError { error });
-    })
-    .forward(out)
+    .filter_map(move |frame| future::ready(match frame {
+        Ok(frame) => {
+            let host = host.clone();
+            source.build_event(frame, host).map(Ok)
+        }
+        Err(_) => {
+            // Drop error, or?
+            Some(Err(()))
+        }
```

Comment:
I don't think we should ever just drop a specific error, but it's not obvious to me what should happen here.

Reply:
Added `warn!` and return `None` instead `Some(Err(()))`